### ISSUE TITLE
wxReactor timer improvements (?)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7198]
|<img alt="xjeethux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| xjeethux reported|
|-|-|
|Trac ID|trac#474|
|Type|enhancement|
|Created|2004-01-15 01:44:21Z|

Attachments:

* [wxreactor.diff.bz2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/162/1625434cebb5e5c4cf7cac668ab24444f9ec05b7/752126f6164442352e3cf87ce18961100fb860a1.bz2) (3114 bytes) - added by xjeethux on 2004-01-15 01:44:20Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__474 474
type__enhancement enhancement
reporter__xjeethux xjeethux
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__invalid invalid
component__core core
keywords__ 
time__1074131061000000 1074131061000000
changetime__1470661848397023 1470661848397023
version__ 
owner__ 
cc__itamarst cc__xjeethux
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="xjeethux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|xjeethux commented|
|-|-|

```
#!html
<pre>
A hack of wxReactor, which uses another wxTimer to implement reactor.callLater
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
what's up with this issue?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
It's in "maybe we should deprecate wxreactor since it doesn't work" limbo.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="xjeethux's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|xjeethux commented|
|-|-|

```
#!html
<pre>
I've seen that wxTimer is wildly inaccurate on win32. But I guess if we can
convince Robin to include a wxSocket wrapper, we can get something workable.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Get him to include a wxSocket wrapper we can use and we might actually have
working wx integration, yeah :)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

You may want to update this patch for the latest wxreactor in 2.5. Also, you say wxCallAfter doesn't always work, got any examples?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

In the future, please don't submit patches that delete and re-add entire files with a different line ending convention.  It makes it very hard to see what significant change is being made.

It seems the patch is changing `ReactorTimer` to do a lot more things.  Here's the relevant code:

```
+class ReactorTimer1(wxTimer):
+    """A Timer for implementing the IReactorTime API in wxReactor"""
+    
+    def __init__(self, reactor): 
+        wxTimer.__init__(self)   
+        self.reactor = reactor   
+
+    def Start(self,t,o):
+        """Remaps t in secs to t in ms
+        """
+        if t==0:
+            t=1 
+        t = int(t*1000)
+        try:
+            wxTimer.Start(self,t,o)
+        except:
+            print 'Exception Caught t=%d'%(t)
+            raise
+ 
+    def Notify(self): 
+        """Called every timer interval"""
+        now = seconds()
+        pcl = self.reactor._pendingTimedCalls
+        while pcl and \
+              (pcl[-1].time <= now):
+            call = pcl.pop()
+            try:
+                call.called = 1
+                call.func(*call.args, **call.kw)
+            except:
+                log.deferr()
+                if hasattr(call, "creator"):
+                    e = "\n"
+                    e += " C: previous exception occurred in " + \
+                         "a DelayedCall created here:\n"
+                    e += " C:"
+                    e += "".join(call.creator).rstrip().replace("\n","\n C:")
+                    e += "\n"
+                    log.msg(e)
+        if self.reactor._pendingTimedCalls:
+            t = self.reactor._pendingTimedCalls[-1].time - now
+            self.Start(t,wxTIMER_ONE_SHOT)
+        elif not self.reactor.running:
+            # Break the cycle (if it does exist)
+            del self.reactor
```

The code seems to be doing two things:

  1. Multiple the sleep time by 1000.  I don't understand the reason for this.  It appears to simply be wrong.  It will result in every sleep lasting 1000 times as long as it is supposed to.  Perhaps the intent is that the overridden Start will only be called by code that thinks it should pass in values which are 1/1000th as large as they should be.  This is a poor API decision, if so.
  1. Re-implement DelayedCall processing.  This is worse than pointless - this code should not be duplicated in the wxreactor.  It doesn't seem to have any significant differences from the original implementation of this code.

So if the patch fixes a problem, I don't know what it is.  In the mean time, wxreactor has been rewritten significantly at least twice, so whatever problem originally existed may have been removed.  Further, wxPython itself has changed significantly, so whatever problem originally existed may have been eliminated on that end.

Please, re-open with a description of the problem to be solved if the problem still exists.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|GitHub commented|
|-|-|

In changeset a92af71eb405a57be477d6970ed91e256c13fbbb
```
#!CommitTicketReference repository="" revision="a92af71eb405a57be477d6970ed91e256c13fbbb"
Merge pull request #474 from twisted/8717-rodrigc-protocols-py3

Author: rodrigc
Reviewer: acabhishek942
Fixes: #8717

Fix twisted.test.test_protocols on Python 3
```
